File: cpp/RemoveDuplicatesFromSortedListII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        ListNode *dummy = new ListNode(0);
        dummy->next = head;
        ListNode *p = dummy;
        while (p && p->next) {
            ListNode *slow = p->next;
            ListNode *fast = slow;
            while (fast->next && fast->next->val == slow->val) {
                fast = fast->next;
            }
            if (fast != slow) {
                fast = fast->next;
            }x
            p->next = fast;
            if (fast == slow) {
                p = p->next;
            }
        }
        return dummy->next;
    }
};

File: java/RemoveDuplicatesFromSortedListII.java
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode skip = pre.next;
        while (skip != null && skip.next != null) {
            while (skip.next != null && skip.next.val == skip.val) {
                skip = skip.next;
            }
            
            if (pre.next != skip) {
                pre.next = skip.next;
                skip.next = null;
                skip = pre.next;
            } else {
                pre = pre.next;
                skip = skip.next;
            }
        }
        
        return dummy.next;
    }
}
